### ISSUE TITLE
prefer-const rule false positive

### ISSUE THREAD

**Issue by egoroof** [eslint/eslint#5074]
Version 2.0.0.beta.2 also checked with latest master https://github.com/eslint/eslint/commit/6433b7b4372468d3470d32f6c2285fa9f06e24fa.
Code is:
```js
let id;

function foo() {
    if (typeof id !== 'undefined') {
        return;
    }
    id = setInterval(() => {
        // some stuff
    }, 250);
}

foo();
```
The output is `1:5  error  'id' is never modified, use 'const' instead  prefer-const`.
But `id` is modified and declaration with `let` shouldn't alert error in this case.
(moved from #5024 as it is not the same)

**Comment by eslintbot** [comment]
@egoroof Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by lo1tuma** [comment]
Thanks for the issue @egoroof. Can you please provide all the information requested by @eslintbot?

Especially interesting would be the `sourceType` you have configured.

**Comment by egoroof** [comment]
Here is my config:

```js
module.exports = {
    "parser": "babel-eslint",
    "env": {
        "es6": true,
        "browser": true,
        "commonjs": true
    },
    "globals": {
        "chrome": false
    },
    "rules": { // http://eslint.org/docs/rules/

        // possible errors
        "comma-dangle": 2, // disallow or enforce trailing commas
        "no-cond-assign": 2, // disallow assignment in conditional expressions
        "no-console": 0, // disallow use of console in the node environment
        "no-constant-condition": 2, // disallow use of constant expressions in conditions
        "no-control-regex": 2, // disallow control characters in regular expressions
        "no-debugger": 2, // disallow use of debugger
        "no-dupe-args": 2, // disallow duplicate arguments in functions
        "no-dupe-keys": 2, // disallow duplicate keys when creating object literals
        "no-duplicate-case": 2, // disallow a duplicate case label.
        "no-empty-character-class": 2, // disallow the use of empty character classes in regular expressions
        "no-empty": 2, // disallow empty statements
        "no-ex-assign": 2, // disallow assigning to the exception in a catch block
        "no-extra-boolean-cast": 2, // disallow double-negation boolean casts in a boolean context
        "no-extra-parens": 2, // disallow unnecessary parentheses
        "no-extra-semi": 2, // disallow unnecessary semicolons
        "no-func-assign": 2, // disallow overwriting functions written as function declarations
        "no-inner-declarations": 2, // disallow function or variable declarations in nested blocks
        "no-invalid-regexp": 2, // disallow invalid regular expression strings in the RegExp constructor
        "no-irregular-whitespace": 2, // disallow irregular whitespace outside of strings and comments
        "no-negated-in-lhs": 2, // disallow negation of the left operand of an in expression
        "no-obj-calls": 2, // disallow the use of object properties of the global object (Math and JSON) as functions
        "no-regex-spaces": 2, // disallow multiple spaces in a regular expression literal
        "no-sparse-arrays": 2, // disallow sparse arrays
        "no-unexpected-multiline": 2, // Avoid code that looks like two expressions but is actually one
        "no-unreachable": 2, // disallow unreachable statements after a return, throw, continue, or break statement
        "use-isnan": 2, // disallow comparisons with the value NaN
        "valid-jsdoc": 2, // Ensure JSDoc comments are valid
        "valid-typeof": 2, // Ensure that the results of typeof are compared against a valid string

        // best practices
        "accessor-pairs": 2, // Enforces getter/setter pairs in objects
        "block-scoped-var": 2, // treat var statements as if they were block scoped
        "complexity": 2, // specify the maximum cyclomatic complexity allowed in a program
        "consistent-return": 2, // require return statements to either always or never specify values
        "curly": 2, // specify curly brace conventions for all control statements
        "default-case": 0, // require default case in switch statements
        "dot-location": [2, "property"], // enforces consistent newlines before or after dots
        "dot-notation": 2, // encourages use of dot notation whenever possible
        "eqeqeq": 2, // require the use of === and !==
        "guard-for-in": 2, // make sure for-in loops have an if statement
        "no-alert": 2, // disallow the use of alert, confirm, and prompt
        "no-caller": 2, // disallow use of arguments.caller or arguments.callee
        "no-case-declarations": 2, // disallow lexical declarations in case clauses
        "no-div-regex": 2, // disallow division operators explicitly at beginning of regular expression
        "no-else-return": 2, // disallow else after a return in an if
        "no-empty-label": 2, // disallow use of labels for anything other than loops and switches
        "no-empty-pattern": 2, // disallow use of empty destructuring patterns
        "no-eq-null": 2, // disallow comparisons to null without a type-checking operator
        "no-eval": 2, // disallow use of eval()
        "no-extend-native": 2, // disallow adding to native types
        "no-extra-bind": 2, // disallow unnecessary function binding
        "no-fallthrough": 2, // disallow fallthrough of case statements
        "no-floating-decimal": 2, // disallow the use of leading or trailing decimal points in numeric literals
        "no-implicit-coercion": 2, // disallow the type conversions with shorter notations
        "no-implied-eval": 2, // disallow use of eval()-like methods
        "no-invalid-this": 2, // disallow this keywords outside of classes or class-like objects
        "no-iterator": 2, // disallow usage of __iterator__ property
        "no-labels": 2, // disallow use of labeled statements
        "no-lone-blocks": 2, // disallow unnecessary nested blocks
        "no-loop-func": 2, // disallow creation of functions within loops
        "no-magic-numbers": 0, // disallow the use of magic numbers
        "no-multi-spaces": 2, // disallow use of multiple spaces
        "no-multi-str": 2, // disallow use of multiline strings
        "no-native-reassign": 2, // disallow reassignments of native objects
        "no-new-func": 2, // disallow use of new operator for Function object
        "no-new-wrappers": 2, // disallows creating new instances of String,Number, and Boolean
        "no-new": 2, // disallow use of the new operator when not part of an assignment or comparison
        "no-octal-escape": 2, // disallow use of octal escape sequences in string literals, such as var foo = "Copyright \251";
        "no-octal": 2, // disallow use of octal literals
        "no-param-reassign": 2, // disallow reassignment of function parameters
        "no-process-env": 2, // disallow use of process.env
        "no-proto": 2, // disallow usage of __proto__ property
        "no-redeclare": 2, // disallow declaring the same variable more than once
        "no-return-assign": 2, // disallow use of assignment in return statement
        "no-script-url": 2, // disallow use of javascript: urls
        "no-self-compare": 2, // disallow comparisons where both sides are exactly the same
        "no-sequences": 2, // disallow use of the comma operator
        "no-throw-literal": 2, // restrict what can be thrown as an exception
        "no-unused-expressions": 2, // disallow usage of expressions in statement position
        "no-useless-call": 2, // disallow unnecessary .call() and .apply()
        "no-useless-concat": 2, // disallow unnecessary concatenation of literals or template literals
        "no-void": 2, // disallow use of the void operator
        "no-warning-comments": 1, // disallow usage of configurable warning terms in comments e.g. TODO or FIXME
        "no-with": 2, // disallow use of the with statement
        "radix": 2, // require use of the second argument for parseInt()
        "vars-on-top": 2, // require declaration of all vars at the top of their containing scope
        "wrap-iife": 2, // require immediate function invocation to be wrapped in parentheses
        "yoda": 2, // require or disallow Yoda conditions

        // strict mode
        "strict": 0, // controls location of Use Strict Directives

        // variables
        "init-declarations": 0, // enforce or disallow variable initializations at definition
        "no-catch-shadow": 2, // disallow the catch clause parameter name being the same as a variable in the outer scope
        "no-delete-var": 2, // disallow deletion of variables
        "no-label-var": 2, // disallow labels that share a name with a variable
        "no-shadow-restricted-names": 2, // disallow shadowing of names such as arguments
        "no-shadow": 2, // disallow declaration of variables already declared in the outer scope
        "no-undef-init": 2, // disallow use of undefined when initializing variables
        "no-undef": 2, // disallow use of undeclared variables unless mentioned in a /*global */ block
        "no-undefined": 2, // disallow use of undefined variable
        "no-unused-vars": 2, // disallow declaration of variables that are not used in the code
        "no-use-before-define": 2, // disallow use of variables before they are defined

        // nodejs and commonjs
        "callback-return": 2, // enforce return after a callback
        "global-require": 2, // enforce require() on top-level module scope
        "handle-callback-err": 2, // enforce error handling in callbacks
        "no-mixed-requires": 2, // disallow mixing regular variable and require declarations
        "no-new-require": 2, // disallow use of new operator with the require function
        "no-path-concat": 2, // disallow string concatenation with __dirname and __filename
        "no-process-exit": 2, // disallow process.exit()
        "no-restricted-modules": 2, // restrict usage of specified node modules
        "no-sync": 2, // disallow use of synchronous methods

        // stylictic issues
        "array-bracket-spacing": 2, // enforce spacing inside array brackets
        "block-spacing": 2, // disallow or enforce spaces inside of single line blocks
        "brace-style": 2, // enforce one true brace style
        "camelcase": 2, // require camel case names
        "comma-spacing": 2, // enforce spacing before and after comma
        "comma-style": 2, // enforce one true comma style
        "computed-property-spacing": 2, // require or disallow padding inside computed properties
        "consistent-this": 2, // enforce consistent naming when capturing the current execution context
        "eol-last": 2, // enforce newline at the end of file, with no multiple empty lines
        "func-names": 0, // require function expressions to have a name
        "func-style": [1, "declaration"], // enforce use of function declarations or expressions
        "id-length": 0, // this option enforces minimum and maximum identifier lengths (variable names, property names etc.)
        "id-match": 2, // require identifiers to match the provided regular expression
        "indent": 1, // specify tab or space width for your code
        "jsx-quotes": 2, // specify whether double or single quotes should be used in JSX attributes
        "key-spacing": 2, // enforce spacing between keys and values in object literal properties
        "linebreak-style": [2, "windows"], // disallow mixed 'LF' and 'CRLF' as linebreaks
        "lines-around-comment": 2, // enforce empty lines around comments
        "max-depth": 2, // specify the maximum depth that blocks can be nested
        "max-len": [2, 120], // specify the maximum length of a line in your program
        "max-nested-callbacks": 2, // specify the maximum depth callbacks can be nested
        "max-params": [2, 5], // limits the number of parameters that can be used in the function declaration.
        "max-statements": 2, // specify the maximum number of statement allowed in a function
        "new-cap": 2, // require a capital letter for constructors
        "new-parens": 2, // disallow the omission of parentheses when invoking a constructor with no arguments
        "newline-after-var": 2, // require or disallow an empty newline after variable declarations
        "no-array-constructor": 2, // disallow use of the Array constructor
        "no-bitwise": 2, // disallow use of bitwise operators
        "no-continue": 2, // disallow use of the continue statement
        "no-inline-comments": 0, // disallow comments inline after code
        "no-lonely-if": 2, // disallow if as the only statement in an else block
        "no-mixed-spaces-and-tabs": 2, // disallow mixed spaces and tabs for indentation
        "no-multiple-empty-lines": 2, // disallow multiple empty lines
        "no-negated-condition": 2, // disallow negated conditions
        "no-nested-ternary": 2, // disallow nested ternary expressions
        "no-new-object": 2, // disallow the use of the Object constructor
        "no-plusplus": 0, // disallow use of unary operators, ++ and --
        "no-restricted-syntax": 2, // disallow use of certain syntax in code
        "no-spaced-func": 2, // disallow space between function identifier and application
        "no-ternary": 0, // disallow the use of ternary operators
        "no-trailing-spaces": 2, // disallow trailing whitespace at the end of lines
        "no-underscore-dangle": 2, // disallow dangling underscores in identifiers
        "no-unneeded-ternary": 2, // disallow the use of ternary operators when a simpler alternative exists
        "object-curly-spacing": 2, // require or disallow padding inside curly braces
        "one-var": 0, // require or disallow one variable declaration per function
        "operator-assignment": 2, // require assignment operator shorthand where possible or prohibit it entirely
        "operator-linebreak": 2, // enforce operators to be placed before or after line breaks
        "padded-blocks": 0, // enforce padding within blocks
        "quote-props": 0, // require quotes around object literal property names
        "quotes": [1, "single"], // specify whether backticks, double or single quotes should be used
        "require-jsdoc": 0, // Require JSDoc comment
        "semi-spacing": 2, // enforce spacing before and after semicolons
        "semi": 2, // require or disallow use of semicolons instead of ASI
        "sort-vars": 2, // sort variables within the same declaration block
        "space-after-keywords": 2, // require a space after certain keywords
        "space-before-blocks": 2, // require or disallow a space before blocks
        "space-before-function-paren": 0, // require or disallow a space before function opening parenthesis
        "space-before-keywords": 2, // require a space before certain keywords
        "space-in-parens": 2, // require or disallow spaces inside parentheses
        "space-infix-ops": 2, // require spaces around operators
        "space-return-throw-case": 2, // require a space after return, throw, and case
        "space-unary-ops": 2, // require or disallow spaces before/after unary operators
        "spaced-comment": 2, // require or disallow a space immediately following the // or /* in a comment
        "wrap-regex": 2, // require regex literals to be wrapped in parentheses

        // ecmascript 6
        "arrow-body-style": 2, // require braces in arrow function body
        "arrow-parens": 2, // require parens in arrow function arguments
        "arrow-spacing": 2, // require space before/after arrow function's arrow 
        "constructor-super": 2, // verify calls of super() in constructors
        "generator-star-spacing": 1, // enforce spacing around the * in generator functions
        "no-arrow-condition": 2, // disallow arrow functions where a condition is expected
        "no-class-assign": 2, // disallow modifying variables of class declarations
        "no-const-assign": 2, // disallow modifying variables that are declared using const
        "no-dupe-class-members": 2, // disallow duplicate name in class members
        "no-this-before-super": 2, // disallow use of this/super before calling super() in constructors.
        "no-var": 2, // require let or const instead of var
        "object-shorthand": 2, // require method and property shorthand syntax for object literals
        "prefer-arrow-callback": 2, // suggest using arrow functions as callbacks
        "prefer-const": 2, // suggest using const declaration for variables that are never modified after declared
        "prefer-reflect": 0, // suggest using Reflect methods where applicable
        "prefer-spread": 2, // suggest using the spread operator instead of .apply().
        "prefer-template": 2, // suggest using template literals instead of strings concatenation
        "require-yield": 2 // disallow generator functions that do not have yield
    }
};
```

What is `sourceType`?

**Comment by lo1tuma** [comment]
`sourceType` allows you to specify whether you are using ES2015 modules or not, which is important to determine if the topmost scope is the global scope or not (see [migration guide](https://github.com/eslint/eslint/blob/master/docs/user-guide/migrating-to-2.0.0.md#language-options)).

Seems like you are not using ES2015 modules, so I would agree that in this case your example is a false positive, as the `id` is a global variable and could be used somewhere.

**Comment by nzakas** [comment]
This one may be tricky to address. I'm not sure that there's a clear signal this shouldn't be a const. Flagging as a bug, I just don't know how to fix it.

**Comment by mysticatea** [comment]
I will fix this.
My idea is that this rule checks there are not ReadReference before the WriteReference.

**Comment by nzakas** [comment]
Sounds as good as anything I can come up with :)

**Comment by lo1tuma** [comment]
I guess we should also never flag if the let declaration was made in the global scope, right?

**Comment by nzakas** [comment]
@lo1tuma can you explain why?

**Comment by lo1tuma** [comment]
Because we don't really know if it gets read somewhere.

Example:

```js
let foo;

function bar() {
    foo = 'hello world';
    console.log(foo);
};

function baz() {
    console.info(window.foo);
}
```

This example is reported as a problem:

```
1:5  error  'foo' is never modified, use 'const' instead  prefer-const
```

**Comment by mysticatea** [comment]
`let foo` does not create the `window.foo` property :)

**Comment by ilyavolodin** [comment]
@mysticatea Even if this is not a module/globalReturn/node extra scope?

**Comment by lo1tuma** [comment]
@mysticatea you are right, I've tried the example in the latest Firefox version, but it still doesn't fully support `let`.

Here is another scenario, having a webpage loading two scripts:

```html
<script src='/foo.js'></script>
<script src='/bar.js'></script>
```

where `foo.js` would declare `let something;` in the global scope and `bar.js` tries to read the global by `console.log(something)`.

**Comment by mysticatea** [comment]
@ilyavolodin Yes. The block bindings are not linked to the global object's properties. Those just shadow the global object's properties. The spec is:

- http://tc39.github.io/ecma262/#table-19 - `CreateGlobalVarBinding` and `CreateGlobalFunctionBinding` mention the global object's properties.
- http://tc39.github.io/ecma262/#sec-globaldeclarationinstantiation - Above operations are not used for the LexicallyScopedDeclarations.

@lo1tuma Oh, right.
And if both file has `let something`, the `bar.js`'s raised a ReferenceError...
We can check `/*exported foo*/`, I thought.

**Comment by egoroof** [comment]
Good job! Thanks for support! :+1:

**Comment by flekschas** [comment]
Unfortunately, I believe the problem still persists. I am able to reproduce the error with the following code:

> test.js

```javascript
class Test {
  static test(d3Selection, callback) {
    let clientRect;

    d3Selection.each(function assessClientRect() {
      clientRect = this.getBoundingClientRect();
      callback(clientRect);
    });
  }
}
export default Test;

```

ESLint version `2.2.0`, also tested with 8ca6bedac12155891d5dca556498ce12a035eccb.

Here's the output:
```bash
$ eslint -v
v2.2.0
$ eslint /where/ever/test.js

/where/ever/test.js
  3:9  error  'clientRect' is never modified, use 'const' instead  prefer-const

✖ 1 problem (1 error, 0 warnings)
```

Here's my `.eslintrc`:
```JSON
{
  env: {
    es6: true
  },
  parserOptions: {
    sourceType: "module"
  },
  rules: {
    prefer-const: 2
  }
}
```

Obviously `clientRect` is assigned the current element's client rectangle on every iteration, so ESLint shouldn't advocate _const_ for `clientRect`.

PS: I wasn't able to reproduce the error with a standard `forEach`.

PPS: Just in case somebody is wondering: `d3Selection` is a D3.js selection of some DOM elements and `each` simply loops over them.

**Comment by mysticatea** [comment]
@flekschas Thank you for reporting.

Actually, this is intentional behavior for now.
You can modify the code with:

```js
class Test {
  static test(d3Selection, callback) {
    d3Selection.each(function assessClientRect() {
      const clientRect = this.getBoundingClientRect();
      callback(clientRect);
    });
  }
}
export default Test;
```


Please see: #5251, #5283, and #5284.
Especially #5284, we have a plan to remove this behavior from `prefer-const`.

**Comment by flekschas** [comment]
@mysticatea Thanks for the clarification and the proposed solution. I understand the intention but the current behavior still feels like a bug to me rather than a feature. For example, initially assigning `undefined` to `clientRect` avoids throw a warning:

```javascript
class Test {
  static test(d3Selection, callback) {
    let clientRect = undefined;

    d3Selection.forEach(function assessClientRect() {
      clientRect = this.getBoundingClientRect();
      callback(clientRect);
    });
  }
}
export default Test;
```

Logically the code above is identical to my first example so shouldn't ESLint either complain in both cases or in none? Or am I missing something?

**Comment by mysticatea** [comment]
@flekschas Please see also:

> Especially #5284, we have a plan to remove this behavior from `prefer-const`.
